src/app/qgisapp.cpp

```diff
@@ -8003,6 +8003,32 @@ void QgisApp::saveEdits( QgsMapLayer *layer, bool leaveEditable, bool triggerRep
   if ( vlayer == activeLayer() )
     mSaveRollbackInProgress = true;
 
+  // wait for end of rendering to avoid lock in case of
+  // splilte and gpkg fix #15498
+  // Wait only if the current layer is in the rendering pool
+  if ( mMapCanvas->mapSettings().layers().contains( vlayer->id() ) )
+  {
+    if ( vlayer->dataProvider()->storageType().count( "SQLite" ) )
```

Comment:
Either the above comment or this test are incorrect. The GPKG driver will identify itself as GPKG and not SQLite. Actually perhaps the issue doesn't occur with GPKG since we turn WAL on, and normally in WAL mode, readers and writers shouldn't conflict.
And shouldn't that also affect the QGIS own spatialite provider ?

Reply:
sorry for my late, I was sick... now checking the correct storageType returned by:
1) plain SQlite
2) spatialite (I remeber is "SQLite with spatialite extension")
3) gpkg

Reply:
1) plain SQLite => "SQlite"
2) external splite => "SQlite"
3) empty splite created in qgis => "SQLite database with SpatiaLite extension"
4) demo gpkg => "GPKG"
=> adapting PR